libs/scalalib/src/mill/javalib/pmd/PmdModule.scala

```diff
@@ -0,0 +1,196 @@
+package mill.javalib
+package pmd
+
+import mainargs.Flag
+import mill.constants.OutFiles.out
+import mill.define.*
+import mill.scalalib.{JavaHomeModule, OfflineSupportModule}
+import mill.util.Jvm
+
+/**
+ * Adds support for analyzing sources with [[https://docs.pmd-code.org/latest/index.html PMD]]
+ * (Programming Mistake Detector).
+ */
+@mill.api.experimental
+trait PmdModule extends CoursierModule, OfflineSupportModule {
+
+  /**
+   * [[https://docs.pmd-code.org/latest/pmd_userdocs_making_rulesets.html Rulesets]] for analysis.
+   * Defaults to XML files, with name prefix ''ruleset'', in one of [[moduleDir]] or ''workspace''.
+   */
+  def pmdRulesets: Task[Seq[PathRef]] = Task {
+    BuildCtx.withFilesystemCheckerDisabled {
+      def isRulesetFile(name: String) = name.startsWith("ruleset") && name.endsWith(".xml")
+      def rulesetsIn(root: os.Path) =
+        val rulesets = os.list.stream(root)
+          .collect:
+            case path if os.isFile(path) && isRulesetFile(path.last) => PathRef(path)
+          .toSeq
+        Option.when(rulesets.nonEmpty)(rulesets)
+
+      rulesetsIn(moduleDir)
+        .orElse(rulesetsIn(Task.ctx().workspace))
+        .getOrElse {
+          Task.log.warn("no rulesets found, must be overridden or specified inline with -R option")
+          Seq()
+        }
+    }
+  }
+
+  /**
+   * The specific [[https://docs.pmd-code.org/latest/pmd_userdocs_cli_reference.html#supported-languages language]]
+   * and version to use when parsing source code for a given language.
+   */
+  def pmdUseVersion: Task[Map[String, String]] = Task(Map.empty[String, String])
+
+  /**
+   * File to which output report is written. If empty, the report is rendered to standard output.
+   */
+  def pmdReportFile: Task[Option[PathRef]] = Task(Option.empty[PathRef])
+
+  /**
+   * Files/folders to analyze. Defaults to
+   *  - `sources`, for a [[JavaModule]]
+   *  - folders under [[moduleDir]] with name prefix ''src'', otherwise
+   */
+  def pmdInputs: Task[Seq[PathRef]] = this match
+    case self: JavaModule => Task(self.sources())
+    case _ => Task {
+        BuildCtx.withFilesystemCheckerDisabled {
+          os.list.stream(moduleDir)
+            .collect:
+              case path
+                  if os.isDir(path) && path.relativeTo(moduleDir).segments.head.startsWith("src") =>
+                PathRef(path)
+            .toSeq
+        }
+      }
```

Comment:
This should be a `Task.Sources` to account for changed files.

Reply:
Done.